Refactor config

alanoe opened the pull request:
- Make options names consistent
- Sync options in config file and command line

Commit:
Move release notes specific options to a separate block in config.yaml

Commit:
Remove branch common option from config.yaml

This default branch was only used in the upgrade-version subcommand,
for which backwards compatibility is not needed. All current package
YAML configuration files have an explicit branch defined, so this
option is not needed anymore.

Commit:
Rename repositories_path attrib in config.yaml to packages_repos_target_path

Commit:
Rename packages metadata repository attributes in config file and command line

New attributes names are more clearer and more consistent with other attrib names.

Commit:
Make subcommand names equal to config node names

Commit:
Rename default block in config.yaml to common

Commit:
Move discover_packages() method to outside config.py

config.py should handle only parsing command line and config YAML file,
nothing else.

Commit:
Create packages_metadata_repo_target_path and http_proxy params in config.yaml

Commit:
Improve build-package and release-notes subcommands names in host_os.py command line

- Rename build-package to build-packages. As subcommand supports
  building of N packages, new name makes more sense
- Rename release-notes to build-release-notes. New name is more
  consistent with other subcommands names such as build-packages
  and build-iso.

Commit:
Create build_packages block in config.yaml

Also, remove packages options where they are never used:
build-release-notes and upgrade-versions subcommands.

Commit:
Synchronize mock and push repo arguments in config.yaml and command line parser

Commit:
Make ISO creation get distro version specific Mock config file

This makes ISO creation consistent with packages creation.

Commit:
Rename package_group_list attrib in config.yaml to automated_install_packages_groups

New name is more intuitive.

Also, renamed related variables in ISO creator.

Commit:
Rename ISO creation attributes in config.yaml to make the names more intuitive

Commit:
Rename ISO builder Python module, class and variables names

"Spin" is Pungi terminology for building a custom ISO. We should not
propagate this anti-intuitive terminology in our code.

Commit:
Synchronize ISO build attributes in config file and host_os.py command line

Commit:
Rename subparsers manager from subparser to subparsers to make code easier to understand

Also add subcommands param to _add_subparsers() and _setup_config_parser() instead of
using the variable defined outside the class directly. That restricts knowledge of
external variables to a single method in the class.

Also add doc to related methods.

Commit:
Reorganize config parser Python module

- Remove _add_subparsers(). This method is small, the calling method
  is small and this method is used only once, so there is no need to
  split it in a separate method.

- Improve methods documentation

- Improve methods names

Commit:
Rename upgrade-versions command line subcommand to update-versions

'update' is the terminology used in the rest of the UI and code.

Also, rename upgrade_versions attribute in config file to update_versions.

Commit:
Move selection of package builder and its conf file to BuildManager

Before, the selection of the package builder and its configuration file was
done by the Distribution class. This patch moves this reponsibility to the
facade BuildManager class and restricts the Distribution class responsibilities
to keeping the Linux distribution name, version, arch, endianness and checking
if it is supported by the build system. The purpose is to make the class design
simpler to understand.

Also, move PackageBuilder class to a separate Python module.

Commit:
Add log_size and verbose options to config.yaml

These options already exists in command line, but not in config file.

olavphibm commented on config.yaml:
```diff
@@ -1,8 +1,8 @@
  packages_repos_target_path: "/var/lib/host-os/repositories"
- build_versions_repository_url: "https://github.com/open-power-host-os/versions.git"
- build_version: 'master'
+ packages_metadata_repo_url: "https://github.com/open-power-host-os/versions.git"
```
Since this is exposed to the user and a common option to use, I'd prefer to use a more relatable name, like `packages_versions_repo_url`.

farosas replied:
The real problem here is the `build` in the name of the key, which is very confusing. Other than that I think it's a matter of defining if we want to be as generic as possible and lose usability or if we don't care having some 'domain language' for the user to learn. In this particular case, I think the latter is simpler and it will save us time.

olavphibm replied:
I think "packages metadata" is too generic. The new user will not have any clue what is this metadata. By having "versions" in the name, he at least knows what repo it refers to. It also makes explicit *only* the versions repo (or a fork) can be used here.
Having to write "packages metadata" in every variable and documentation, in my opinion, just adds extra words and complexity with no gain of value.

alanoe replied:
OH> I think "packages metadata" is too generic.

But that's exactly what this git repository holds, packages metadata! Any other term I can think of is imprecise and unclear. You think "versions" is a better name? "versions" is a confusing name because it seems like this git repo keeps only versions number and it keeps much more than that. "packages" is also a little confusing because people may confuse with the packages source code git repositories (it still is better than 'versions', though). I prefer a longer clearer name.

btw nothing prevents us from renaming the 'versions' git repository to 'packages' or 'packages_metadata'

side-note: "packages metadata" is a term already used in parts of the code, this is one way to make the terminology more consistent.

olavphibm commented on config.yaml:
```diff
@@ -17,21 +27,16 @@ common:
 build_release_notes:
  repo_url: "https://github.com/open-power-host-os/open-power-host-os.github.io.git"
  repo_branch: "master"
+ push_repo_url: "git@github.com:open-power-host-os/open-power-host-os.github.io.git"
```
Do not set this default value. We shouldn't push directly to the main repository (it may happen by accident) and any other user will not have permission to do so. We should make it explicit that this parameter is required and needs to be set up properly.

alanoe replied:
I do not understand why we should not push directly to the main repository. In the case of our team, that's exactly what we want to do. You are afraid that a member of our team will run "host_os.py --build-release-notes" accidentally? If that is a possibility, why forcing him to use "host_os.py --build-release-notes --push-repo-url="git@github.com:open-power-host-os/open-power-host-os.github.io.git"  is safer?

In the case of someone outside our team, he/she will not have permission to do so, as you mentioned, so there is no issue IMHO.

olavphibm replied:
We always push to the main repository through PRs, not directly. We may want to push directly automatically in our weekly builds in the future, but that can be explicitly set in the weekly job.

Someone may be testing the commands and not realize the default is to push to the main repository. Forcing to use `--push-repo-url` makes the user think (or, at least, look) about where he wants to push.

alanoe replied:
Thanks, this is clearer to me now. I thought about using the URL of our bot user git repo as default here. However, as I discussed with Olav today, running update-versions and build-release-versions subcommands won't work out of the box anyway because the user must have the private SSH key of the GitHub user and a SSH config setup in the system running host_os.py. IMHO either command line should parametrize this or at least we should document this so that these commands work in the command line and we do not have a dependency with Jenkins here, but Olav told me this has low priority. I'll just leave the attribute blank for now.

olavphibm commented on config.yaml:
```diff
@@ -80,3 +85,6 @@ build_iso:
         updates: "http://mirror.centos.org/altarch/7/updates/ppc64le/"
         epel: "http://download.fedoraproject.org/pub/epel/7/ppc64le/"
+upgrade_versions:
+ push_repo_url: "git@github.com:open-power-host-os/versions.git"
```
Same as above.

olavphibm commented on lib/config.py:
```diff
@@ -59,9 +59,12 @@
 }
 MOCK_ARGS = {
+    ('--mock-binary',):
+        dict(help='Mock binary path', default='/usr/bin/mock'),
```
There's no need to have this default here, it's already in the yaml file.

alanoe replied:
I prefer to remove all the default values from the command line parser in a separate patch

olavphibm commented on lib/config.py:
```diff
@@ -59,9 +59,12 @@
+    ('--mock-binary',):
+        dict(help='Mock binary path', default='/usr/bin/mock'),
+    ('--mock-config',):
+        dict(help='Mock config file', default=''),
```
Why not omit the `default` parameter and leave the default value `None`?

alanoe replied:
Done

olavphibm commented on lib/iso_spinner.py:
```diff
@@ -114,13 +114,13 @@ def _create_spin_kickstart(self):
+                if hostos_group not in groups:
+                    groups.append(hostos_group)
+            for group in groups:
+                f.write(group + "\n")
```
Regarding commit message:
"Also, implemented a related small variable renaming in package group creation code
in ISO creator."
I'd write something simpler:
"Also, renamed related variables in ISO creator."

alanoe replied:
Done

olavphibm commented on config.yaml:
```diff
@@ -80,10 +80,10 @@ build_iso:
         name: "hostos"
         dir: "/host-os-repo"
-    distro_repo_url:
+    distro_repo_urls:
```
`distro_repos_urls`

alanoe replied:
Done

olavphibm commented on lib/config.py:
```diff
@@ -100,6 +100,25 @@
              default='./result'),
+    ('--iso-name',):
+        dict(help='ISO name.',
+             default='OpenPOWER-Host_OS'),
```
There's no need to have those defaults here. They are already in the config.yaml file.

alanoe replied:
I prefer to remove all the default values from the command line parser in a separate patch

olavphibm commented on lib/config.py:
```diff
@@ -100,6 +100,25 @@
+        dict(help='ISO name.',
+             default='/var/log/host-os/iso.log'),
+    ('--kickstart-file',):
+        dict(help='kickstart file path',
```
Most help texts start with uppercase.

alanoe replied:
Done

olavphibm commented on lib/config.py:
```diff
@@ -188,47 +189,55 @@ def _setup_config_parser_options(self, subcommands):
+        Parse configuration from a configuration file
+
+        Args:
+            config_file_path(str): configuration file path
```
You could mention the format of the configuration file (yaml)

alanoe replied:
Done

olavphibm commented on lib/config.py:
```diff
@@ -188,47 +189,55 @@ def _setup_config_parser_options(self, subcommands):
+            config_file_path(str): configuration file path
+
+        Returns:
+            dict: Configuration options provided by user in configuration file.
```
I'd write: "options set by the configuration file"
The user might not have modified the file.

alanoe replied:
Done

olavphibm commented on lib/config.py:
```diff
@@ -188,47 +189,55 @@ def _setup_config_parser_options(self, subcommands):
+
     def parse(self):
+        """
+        Parse configuration
```
This doesn't say much. What about:
"Parse configuration file and command line arguments"

alanoe replied:
Done

olavphibm commented on lib/build_manager.py:
```diff
@@ -47,6 +49,48 @@ def __call__(self):
+        if distro.lsb_name == "CentOS":
+            mock_config_file = CONF.get('build_packages').get('mock_config').get(
+                distro.lsb_name).get(distro.version)
+            package_builder = mockbuilder.Mock(mock_config_file)
```
I don't know if we should keep this distro specifics here, but this looks like a minor drawback from having the build functions removed from the distro class.

alanoe replied:
There is a 1-N relationship between package builder and distro. Handling this logic in the distro class or a third class are the possible solutions. As the number of distros we'll support is small and the logic to infer the package builder from the distro is very simple, IMHO keeping this logic in a third class (in this case, build manager) is simpler and clearer.

olavphibm commented on config.yaml:
```diff
@@ -1,5 +1,6 @@
 common:
  log_file: "/var/log/host-os/builds.log"
+ log_size: 2097152
```
If you're adding it here, it can be removed from config.py.
You could actually remove all defaults from there since now everything is in the yaml file.

alanoe replied:
I prefer to remove all the default values from the command line parser in a separate patch

alanoe commented on lib/config.py:
```diff
@@ -269,9 +270,11 @@ def parse(self):
-                config['iso'][key] = value
-                args.pop(key)
+            for node in COMMAND_TO_CONFIG_NODE.values():
+                if key in config[node]:
```
This is a leftover from a previous implementation idea, fixing it

alanoe commented on lib/config.py:
```diff
@@ -67,9 +67,9 @@
 }
 RELEASE_NOTES_ARGS = {
-    ('--release-notes-repo-url',):
+    ('--release-notes.repo-url',):
```
Leftover from a previous implementation idea, fixing it

alanoe commented on lib/package.py:
```diff
@@ -138,7 +138,7 @@ def _download_source_code(self):
         self._setup_repository(
             dest=CONF.get('default').get('repositories_path'),
-            branch=CONF.get('default').get('branch'))
+            branch='powerkvm-v3.1.1')
```
Done

alanoe commented on config.yaml:
```diff
@@ -1,6 +1,6 @@
 default:
  log_file: "/var/log/host-os/builds.log"
- repositories_path: "/var/lib/host-os/repositories"
+ packages_repos_target_path: "/var/lib/host-os/repositories"
```
Ok. This directory name makes sense with the current scope of the directory. If the scope is updated with @gutoyr 's patch, he may rename it.

I'm wondering if "package_repo" may be misunderstood with a yum repository...

Commit:
Replace __call__() with a method named with a verb in classes

Defining and using __call__() in classes is a confusing pattern.
It is much more intuitive to use a verb name which repeats the
implicit action in the class name (eg build_manager.build() and
scheduler.schedule() instead of build_manager() and scheduler()).

Commit:
Improve Scheduler class documentation

Commit:
Add option which allows disabling the update of SCM repositories before building

This is useful for developers who want to locally update the repositories
and then rebuild.

Commit:
Support overriding SCM repository URL, commit ID and branch in command line

farosas commented on config.yaml:
```diff
@@ -22,6 +20,9 @@ default:
  force_rebuild: False
+release_notes:
+ release_notes_repo_url: "https://github.com/open-power-host-os/open-power-host-os.github.io.git"
+ release_notes_repo_branch: "master"
```
These names can now be simply `repo_url` and `repo_branch`

alanoe replied:
release-notes subcommand may accept a packages_metadata_repo_url (versions.git) arg and a release_notes_repo_url (open-power-host-os.github.io.git) arg. I thought that calling one of them repo_url would be a little confusing for the command line user, but up to you guys.

farosas commented on lib/config.py:
```diff
@@ -29,7 +29,7 @@
     ('--packages-metadata-repo-branch',):
         dict(help='Packages metadata git repository branch'),
-    ('--build-versions-repo-dir',):
+    ('--packages-metadata-repo-target-path',):
```
This patch could be squashed with the previous one

alanoe replied:
Done

farosas commented on lib/config.py:
```diff
@@ -251,7 +251,7 @@ def parse(self):
         # at the same level of the 'default' node
         COMMAND_TO_CONFIG_NODE = {
-            "build-iso": "iso",
+            "build-iso": "build_iso",
```
We could now simplify this implementation using a list of sub-command names.

alanoe replied:
I'll do it in a separate patch

farosas commented on tests/unit/config_tests.py:
```diff
@@ -17,7 +17,7 @@ class TestConfigParser(unittest.TestCase):
         (['build-packages', '--result-dir=foo'], 'result_dir', 'foo'),
         (['build-packages', '--packages-repos-target-path=foo'], 'packages_repos_target_path', 'foo'),
-        (['build-packages', '--keep-builddir'], 'keep_builddir', True),
+        (['build-packages', '--keep-build-dir'], 'keep_builddir', True),
```
The second element of the tuple should change too.

alanoe replied:
Done

Commit:
Remove default values from host_os.py's command line arguments

farosas commented on tools/build_release_notes.py:
```diff
@@ -144,8 +144,7 @@ def push_website_head_commit(
 
-    packages_names = (CONF.get('common').get('packages')
-                      or packages_manager.discover_packages())
+    packages_names = packages_manager.discover_packages()
```
What do you mean with 'are never used'? And what about command line usage?

alanoe replied:
We only use this subcommand to create release notes of all packages. What is the usefulness of doing it for a subset?

farosas replied:
I don't know, just thought we would like to keep consistent with build-packages. Ok then.

farosas commented on config.yaml:
```diff
@@ -35,7 +35,9 @@ build_release_notes:
-    mock_config: "./mock_configs/CentOS/7/spin-iso-CentOS-7-ppc64le.cfg"
+    mock_config:
+     CentOS:
+      '7': "./mock_configs/CentOS/7/spin-iso-CentOS-7-ppc64le.cfg"
```
side note: why don't we infer the path from the yaml keys?

alanoe replied:
+1. We can improve it in a separate PR.

farosas commented on lib/config.py:
```diff
@@ -100,6 +100,25 @@
+    ('--log-file',):
+        dict(help='ISO name.',
+             default='/var/log/host-os/iso.log'),
+    ('--kickstart-file',):
```
This name changed.

alanoe replied:
Done

farosas commented on lib/config.py:
```diff
@@ -100,6 +100,25 @@
+        dict(help='Packages groups in yum repository'),
+    ('--automated-install-package-groups',):
+        dict(help='Packages and package groups which are installed using automated installation'),
+    ('--mock_iso_repo.name',):
```
Underscores and dot instead of dashes

alanoe replied:
Done

farosas commented on lib/config.py:
```diff
@@ -100,6 +100,25 @@
+        dict(help='Packages and package groups which are installed using automated installation'),
+    ('--mock_iso_repo.name',):
+        dict(help='Name of the yum repository which will be created with OpenPOWER Host OS packages'),
+    ('--mock_iso_repo.dir',):
```
Same here

alanoe replied:
Done

farosas commented on lib/package_builder.py:
```diff
@@ -34,3 +30,6 @@ def clean(self):
         """
+
+
+
```
whitespace

alanoe replied:
Done

farosas commented on lib/config.py:
```diff
@@ -43,6 +43,9 @@
     ('--packages-repos-target-path', '-R'):
         dict(help='Directory where to clone code repositories',
              default='/var/lib/host-os/repositories'),
+    ('--no-update-packages-repos-target-path',):
```
name does not match the config key/description

alanoe replied:
Done

farosas reviewed (dismissed):
Aside from what's inline, a couple of points: 

There are too many commits in this PR and some of them are not related to the proposed change, this makes it hard to review and know which changes we already agreed on and which are new. I suggest that in the future you try to map PRs to trello tasks to make life easier for everyone.

I would also dedicate a separate PR for patches which are simply renaming things, because they tend to bring a lot of discussion and we don't want them to delay the "feature" patches or compromise the overall review process.

Commit:
Improve merging of config file attributes and command line arguments in a dictionary

This also fixes unit tests which validate default values to not parse only
command line arguments, but config file too.

Commit:
Support YAML config file dict nodes overriding with command line arguments

Command line argument must be passed in JSON format.

Commit:
Remove mock_config attribute from config YAML

It can be inferred from Linux distro name, version, arch
and endianness.

Commit:
Make layout fixes in config.yaml

- Make indentation consistent
- Remove extra whitespaces

olavphibm commented on lib/config.py:
```diff
@@ -158,16 +158,23 @@ def __init__(self):
         self.parser = argparse.ArgumentParser()
         self._CONF = None
-        self._setup_config_parser_options()
+        self._setup_config_parser_options(SUBCOMMANDS)
```
I'd rewrite the commit message to make the main point clearer. It is not the method rename, it is the addition of the parameter. Also, please try to keep the short commit messages short. It is really annoying to read messages like this:
```
Rename subparsers manager from subparser to subparsers to make code e…

…asier to understand
```
It also only appears partially in the commits page. I suggest:
```
Add "subcommands" parameter to ConfigParser methods

This restricts knowledge of external constants to a single method in the class.
Also, rename the "_add_subparser" method and document related methods.
```

alanoe replied:
Done

olavphibm commented on lib/config.py:
```diff
@@ -28,8 +28,7 @@
         dict(help='Directory to clone the packages metadata git repository. '
-             'A subdirectory with the name of the git repository will be created here',
-             default='.'),
+             'A subdirectory with the name of the git repository will be created here'),
```
Commit message: mentions those defaults are already in config.yaml

alanoe commented:
OH> I suggest keeping this PR with commits until "Add log_size and verbose options to config.yaml" and opening new PRs to the newer ones.

Opened:
https://github.com/open-power-host-os/builds/pull/197
https://github.com/open-power-host-os/builds/pull/198
https://github.com/open-power-host-os/builds/pull/199
https://github.com/open-power-host-os/builds/pull/200

I'll open more PRs when those are merged to avoid wasting time with infinite rebases. Could you guys pls review those ASAP?
